### PR TITLE
[model, ops] feat: enable DeepSeek V4 DSA backend

### PR DESCRIPTION
**Pull request by Luosuu**
## Stack
- Base PR: #899 (`codex/deepseek-v4-fp8-dequant`) handles DeepSeek V4 FP8/int8 checkpoint scale dequantization.
- This PR contains only the DSA backend integration on top of #899.

## Summary
- add explicit DeepSeek sparse attention backends for `flashmla_cudnn` and `fa4_cudnn`
- route DSA backend selection through `OpsImplementationConfig`
- wire DeepSeek V4 CSA layers to FlashMLA/FA4 forward with cuDNN FE DSA backward
- keep backend imports scoped to explicit backend modules and avoid fallback behavior

## Tests
- ruff check on DSA touched files
- `tests/ops/test_deepseek_sparse_attention.py -q` passed on B200 worker before splitting
- DeepSeek V4 patchgen check passed on B200 worker before splitting
- local pytest/patchgen are blocked by local torch missing `libnvshmem_host.so.3`

### CODE REVIEW

**Comment by Luosuu** [comment]
Validation update for the latest head (`8f1bec5b`):

- Local checks passed:
  - `uv run --extra gpu --dev patchgen --check`
  - `uv run --frozen pytest -q tests/ops/test_deepseek_sparse_attention.py` (`17 passed`)
  - `ruff check` / `ruff format --check` on touched files
- B200 worker validation passed:
  - backend FlashMLA/cuDNN DSA backward probe with and without sink
  - DeepSeek-V4-Flash real config-shape attention smoke (`hidden_size=4096`, `num_attention_heads=64`, `head_dim=512`, `qk_rope_head_dim=64`, `sliding_window=128`, `index_topk=512`, `compress_rate=4`)
  - observed one `flashmla_cudnn_call`, finite hidden/sink gradients, and `DEEPSEEK_V4_FLASH_DSA_E2E_OK`
- H100 reached the same FlashMLA/cuDNN path, but cuDNN DSA backward compile fails in the upstream sm90 CUTLASS DSL path with `atomicrmw() missing 1 required positional argument: 'res'`. B200/sm100 is the validated path for this PR.

**Comment by Luosuu** [comment]
Validation update for current head (`63fc5c6`):

- Local checks passed:
  - `uv run --extra gpu --dev patchgen --check`
  - `uv run --frozen pytest -q tests/ops/test_deepseek_sparse_attention.py` (`18 passed`)
  - `ruff check` / `ruff format --check` on touched files
- B200 worker validation passed on `NVIDIA B200 (sm100)`, torch `2.11.0+cu130`, Python `3.12`:
  - DeepSeek-V4-Flash real config-shape CSA attention smoke (`hidden_size=4096`, `num_attention_heads=64`, `head_dim=512`, `qk_rope_head_dim=64`, `sliding_window=128`, `index_topk=512`, `compress_rate=4`)
  - observed one `flashmla_cudnn_call`, finite loss, finite hidden/sink gradients, and `DEEPSEEK_V4_FLASH_DSA_E2E_OK`
- I also compared `eager` vs `flashmla_cudnn` on the same real config-shape attention module; both produced finite forward outputs and finite backward gradients.

File: tests/ops/test_deepseek_sparse_attention.py [hunk 1]
```diff
@@ -140,66 +165,104 @@ def fake_flash_mla_sparse_fwd(q, kv, indices, sm_scale, d_v):
...
-    k_pe = torch.empty(1, 4, 1, 64, dtype=torch.bfloat16)
-    kv_cache = torch.empty(1, 4, 1, 512, dtype=torch.bfloat16)
-    q_nope = torch.empty(1, 2, 128, 512, dtype=torch.bfloat16)
-    gather = torch.zeros(1, 2, 64, dtype=torch.int32)
+    top_k_indices = torch.full((1, 3, 3), -1, dtype=torch.long)
+    top_k_indices[0, 2] = torch.tensor([0, 1, -1])
 
-    compatible, reason = dsa.check_flash_mla_sparse_forward_compatible(q_pe, k_pe, kv_cache, q_nope, gather)
+    sparse_indices, topk_length = _deepseek_v4_build_flashmla_indices(
+        None,
```
Comment (FoolPlayer) [inline tests/ops/test_deepseek_sparse_attention.py:306]: Only test attention_mask = None, but in training case it may not be None.

File: veomni/models/transformers/deepseek_v4/generated/patched_modeling_deepseek_v4_gpu.py [hunk 2]
```diff
@@ -821,28 +913,100 @@ def forward(
...
+            if unsupported_reasons:
+                raise ValueError(
+                    "dsa_attention_backend='flashmla_cudnn' is not supported: " + "; ".join(unsupported_reasons)
+                )
+
+            q_bshd = q.transpose(1, 2)
+            kv_bshd = kv.transpose(1, 2)
+            q_pe = q_bshd[..., -rope_dim:]
+            k_pe = kv_bshd[..., -rope_dim:]
+            q_nope = q_bshd.clone()
```
Comment (FoolPlayer) [inline veomni/models/transformers/deepseek_v4/generated/patched_modeling_deepseek_v4_gpu.py]: Maybe clone is not necessary. We can use `q_nope = torch.nn.functional.pad(q_bshd[..., :-rope_dim], (0, rope_dim))`
